--- Ombi/src/Ombi.Store/Ombi.Store.csproj	2025-04-18 11:39:05.869578532 +0300
+++ Ombi/src/Ombi.Store/Ombi.Store.csproj	2025-04-18 11:40:56.804782942 +0300
@@ -19,7 +19,7 @@
     </PackageReference>
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.5" />
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
-    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="8.0.0" />
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="8.0.3" />
     <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="8.0.2" />
     <PackageReference Include="Nito.AsyncEx" Version="5.1.2" />
     <PackageReference Include="Polly" Version="7.2.3" />
