--- chromaprint/src/audio/ffmpeg_audio_reader.h	2025-04-24 02:46:18.128561196 +0300
+++ chromaprint/src/audio/ffmpeg_audio_reader.h	2025-04-24 02:47:17.673474990 +0300
@@ -161,7 +161,18 @@
 	}
 	m_stream_index = ret;
 
-	m_codec_ctx = m_format_ctx->streams[m_stream_index]->codec;
+	m_codec_ctx = avcodec_alloc_context3(codec);
+	if (!m_codec_ctx) {
+		SetError("Could not allocate codec context");
+		return false;
+	}
+
+	ret = avcodec_parameters_to_context(m_codec_ctx, m_format_ctx->streams[m_stream_index]->codecpar);
+	if (ret < 0) {
+		SetError("Could not copy codec parameters to context", ret);
+		return false;
+	}
+
 	m_codec_ctx->request_sample_fmt = AV_SAMPLE_FMT_S16;
 
 	ret = avcodec_open2(m_codec_ctx, codec, nullptr);
